tools/docker-builder/main.go

```diff
@@ -0,0 +1,332 @@
+
+	"istio.io/istio/pilot/pkg/util/sets"
+	testenv "istio.io/istio/pkg/test/env"
+	"istio.io/pkg/log"
+	pkgversion "istio.io/pkg/version"
+)
+
+func main() {
+	rootCmd.Flags().StringVar(&args.Hub, "hub", args.Hub, "docker hub")
+	rootCmd.Flags().StringVar(&args.Tag, "tag", args.Tag, "docker tag")
+
+	rootCmd.Flags().StringVar(&args.BaseVersion, "base-version", args.BaseVersion, "base version to use")
+	rootCmd.Flags().StringVar(&args.ProxyVersion, "proxy-version", args.ProxyVersion, "proxy version to use")
+	rootCmd.Flags().StringVar(&args.IstioVersion, "istio-version", args.IstioVersion, "istio version to use")
+
+	rootCmd.Flags().StringSliceVar(&args.Targets, "targets", args.Targets, "targets to build")
+	rootCmd.Flags().StringSliceVar(&args.Variants, "variants", args.Variants, "variants to build")
+	rootCmd.Flags().StringSliceVar(&args.Architectures, "architecures", args.Architectures, "architectures to build")
+	rootCmd.Flags().BoolVar(&args.Push, "push", args.Push, "push targets to registry")
+	rootCmd.Flags().BoolVar(&args.Save, "save", args.Save, "save targets to tar.gz")
+	rootCmd.Flags().BoolVar(&args.BuildxEnabled, "buildx", args.BuildxEnabled, "use buildx for builds")
+	rootCmd.Flags().BoolVar(&version, "version", version, "show build version")
+
+	if err := rootCmd.Execute(); err != nil {
+		os.Exit(-1)
+	}
+}
+
+var rootCmd = &cobra.Command{
+	Use:   "",
+	Short: "Builds Istio docker images",
+	RunE: func(cmd *cobra.Command, _ []string) error {
+		if version {
+			fmt.Println(pkgversion.Info.GitRevision)
+			os.Exit(0)
+		}
+		log.Infof("Args: %+v", args)
+		if args.Push && args.Save {
+			// TODO(https://github.com/moby/buildkit/issues/1555) support both
+			return fmt.Errorf("--push and --save are mutually exclusive")
+		}
+		tarFiles, err := ConstructBakeFile(args)
+		if err != nil {
+			return err
+		}
+		targets := []string{}
+		for _, t := range args.Targets {
+			targets = append(targets, fmt.Sprintf("docker.%s", t))
+		}
+		if err := RunMake(args, targets...); err != nil {
+			return err
+		}
+		if err := RunBake(args); err != nil {
+			return err
+		}
+		if err := RunSave(args, tarFiles); err != nil {
+			return err
+		}
+
+		return nil
+	},
+}
+
+func RunBake(args Args) error {
+	out := filepath.Join(testenv.IstioOut, "dockerx_build", "docker-bake.json")
+	_ = os.MkdirAll(filepath.Join(testenv.IstioOut, "release", "docker"), 0o755)
+	if err := createBuildxBuilderIfNeeded(args); err != nil {
+		return err
+	}
+	c := VerboseCommand("docker", "buildx", "bake", "-f", out, "all")
+	c.Stdout = os.Stdout
+	return c.Run()
+}
+
+// --save requires a custom builder. Automagically create it if needed
+func createBuildxBuilderIfNeeded(a Args) error {
+	if !a.Save {
+		return nil
+	}
+	if _, f := os.LookupEnv("CI"); !f {
+		// For now only do this for CI, we do not want to mess with users config. And users rarely use --save
+		return nil
+	}
+	return exec.Command("sh", "-c", `
+export DOCKER_CLI_EXPERIMENTAL=enabled
```

Comment:
Doesn’t docker tool have a go client?

Reply:
Yes, but I didn't really want to introduce a new dependency.

We could also use a bunch of exec.Cmd calls for this but its a lot simple to just do this :-)